python/ray/train/batch_predictor.py

```diff
@@ -194,19 +205,78 @@ def __init__(self):
+                        }
+                    elif isinstance(batch_data, np.ndarray):
+                        return batch_data
+                elif batch_format == BatchFormat.PANDAS:
+                    # Select a subset of the pandas columns.
+                    return batch_data[select_columns]
+
+            def _keep_columns_from_input_batch(
+                self,
+                input_batch: DataBatchType,
+                prediction_output_batch: DataBatchType,
+                keep_columns: Optional[List[str]] = None,
+            ):
+                """Return a union of input batch and prediction output batch
+                based on provided columns.
+                """
+                if not keep_columns:
+                    return prediction_output_batch
+                elif keep_columns and (isinstance(input_batch, np.ndarray)):
+                    raise ValueError(
+                        f"Column name(s) {keep_columns} should not be provided "
+                        "for prediction input data type of ``numpy.ndarray``"
+                    )
+                elif batch_format == BatchFormat.NUMPY:
+                    for column in keep_columns:
+                        prediction_output_batch[column] = input_batch[column]
+                    return prediction_output_batch
+                elif batch_format == BatchFormat.PANDAS:
+                    prediction_output_batch[keep_columns] = input_batch[keep_columns]
+                    return prediction_output_batch
+
+            def __call__(self, input_batch: DataBatchType) -> DataBatchType:
+                # TODO (jiaodong): Investigate if there's room to optimize prediction
+                # result joins to minimize GPU <> CPU transfer
+                prediction_batch: DataBatchType = self._select_columns_from_input_batch(
+                    input_batch, select_columns=feature_columns
                 )
-                if keep_columns:
-                    prediction_output[keep_columns] = batch[keep_columns]
-                return convert_batch_type_to_pandas(
-                    prediction_output, cast_tensor_columns
+                prediction_output_batch: DataBatchType = self._predictor.predict(
+                    prediction_batch, batch_format=batch_format, **predict_kwargs
                 )
+                prediction_output_batch: DataBatchType = (
+                    self._keep_columns_from_input_batch(
+                        input_batch, prediction_output_batch, keep_columns=keep_columns
+                    )
+                )
+
+                if batch_format == BatchFormat.NUMPY:
+                    # User code just need to return Numpy format where we will
+                    # internall convert to Arrow format.
+                    return prediction_output_batch
+                else:
+                    return convert_batch_type_to_pandas(
```

Comment:
I don't think this is necessary- we can just return `prediction_output_batch` directly

Reply:
with latest rebased version this shouldn't be needed -- let me try with new test